mmvb_backend/benchmarking_sessions/api/views.py

```diff
@@ -0,0 +1,124 @@
+    BenchmarkingStepStatus,
+)
+from benchmarking_sessions.tasks import run_benchmark
+
+
+# todo: properly document endpoints
+class BenchmarkingSessionViewSet(ModelViewSet):
+    schema = AutoSchema(tags=["BenchmarkingSessions",])
+    serializer_class = BenchmarkingSessionSerializer
+
+    def get_queryset(self):
+        return BenchmarkingSession.objects.all()
+
+    # todo: generate 404
+    @action(methods=["post"], detail=True, url_path="run")
+    def run_benchmark(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
+            BenchmarkingSession, id=kwargs["pk"]
+        )
+        task = run_benchmark.delay(benchmarking_session.id)
+        benchmarking_session.task_id = task.id
+        benchmarking_session.save()
+        return Response(
+            f"/benchmarking-sessions/{benchmarking_session.id}/status",
+            status=status.HTTP_202_ACCEPTED,
+        )
+
+    # todo: document response structure (OpenAPI)
+    @action(methods=["get"], detail=True, url_path="status")
+    def benchmark_status(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
+            BenchmarkingSession, id=kwargs["pk"]
+        )
+        if benchmarking_session.status != BenchmarkingSession.Status.RUNNING:
+            return Response(
+                {"status": benchmarking_session.status},
+                status=status.HTTP_200_OK,
+            )
+
+        result = run_benchmark.AsyncResult(benchmarking_session.task_id)
+        if result.status == "PENDING" or result.info is None:
+            # this is an in-between state where the celery task status is not yet reflected in the main DB,
+            # so either the task is about to begin or just finished -- in either case it's easiest to pretend it is
+            # running
+            return Response(
+                {"status": BenchmarkingSession.Status.INTERMEDIATE},
+                status=status.HTTP_200_OK,
+            )
+
+        for case_responses in result.info["responses"]:
+            for ai_implementation_id, response in case_responses[
+                "responses"
+            ].items():
+                if "value" in response:
+                    del response["value"]
+
+        return Response(
+            {
+                "status": BenchmarkingSession.Status.RUNNING,
+                "report": result.info,
+            },
+            status=status.HTTP_200_OK,
+        )
+
+    # todo: document response structure (OpenAPI)
+    @action(methods=["get"], detail=True, url_path="results")
+    def benchmark_results(self, request, *args, **kwargs):
+        benchmarking_session = get_object_or_404(
+            BenchmarkingSession, id=kwargs["pk"]
+        )
+
+        benchmarking_session_result = BenchmarkingSessionResultsSerializer(
+            benchmarking_session
+        ).data
+
+        # todo: add metrics infrastructure, add more metrics
+        proportion_cases_with_ai_result_counts = {}
```

Comment:
Maybe it is just me but this block looks a little bit hard to read. I would suggest:
```
from collections import defaultdict

proportion_cases_with_ai_result_counts = defaultdict(int)
responses = benchmarking_session_result["responses"]

for case_response in responses:
    case_responses = case_response["responses"]
    for ai_implementation_id, response in case_responses.items():
        step_status_value = BenchmarkingStepStatus.COMPLETED.value
        increment = int(response["status"] == step_status_value)
        proportion_cases_with_ai_result_counts[ai_implementation_id] += increment
```

Reply:
didn't know about `defaultdict`, neat!
i've given it a try to clean up, please have another look.
(same for below)